AK3: add support for resizing logical partitions

capntrips opened the pull request:
This adds `lptools` and a merged `fastboot`/`fastbootd` and utilizes them for logical partitions. I [added](https://github.com/capntrips/vendor_lptools/tree/find-list) an `exists` command to `lptools` (I'll make a PR there if this method is acceptable) and [merged](https://github.com/capntrips/vendor_ufastboot) `fastboot` and `fastbootd` into a single binary with the ability to resize and flash to logical partitions. The latter may also be useful for physical partitions, but I'm not familiar enough with the reasons behind the other flashing methods.

If the heavily modified `fastboot`/`fastbootd` binary is too much, I also [built](https://github.com/capntrips/system_core_fastboot) an untouched `fastboot` that will run on the device and a lightly touched, static `fastbootd` that works on a running system.

**Edit:** I just wanted to note, the reason I went with `fastboot` over a simple resize with `lptools` is the stock `vendor_dlkm` is a sparse image, and `fastboot` supports sparse partitions.

osm0sis commented:
Nice!! Thanks!

I'm on vacation right now so can't review at the moment, but I'll check it out as soon as I can!

Is this all tested and all working? What was your test device? I'm surprised fastboot works on a logical partition device when not in fastbootd!

My understanding of vendor_dlkm was that the Google factory image for it is sparse, but one built from source is not, so you could probably just stick with lptools since no kernel dev would be replacing it with the stock factory image via AnyKernel3, they'd be using their own, which would not be sparse.

capntrips commented:
I tested it on a Pixel 6 (oriole) with Kirisakura 2.1.0, and it is working. I'll take a stab at using `lptools` to do the actual resize with the existing flash methods, assuming a `wc -c` of the image is suitable to determining the partition size.

Today is a travel day for me, so it may be tomorrow before I have chance to look at it.

capntrips committed:
AK3: add support for resizing logical partitions

capntrips commented:
I added some checks and changed the exit status on error for the resize function of `lptools` (see PR [here](https://github.com/phhusson/vendor_lptools/pull/3)).

The `lptools` binary in this commit is fully static (including `libcrypto`) and 32-bit `arm`.

This currently remounts `/vendor_dlkm`, but it might be better to pull the arguments from `fstab` (see [here](https://github.com/aosp-mirror/platform_system_core/blob/android-12.1.0_r4/fastboot/device/flashing.cpp#L59-L73) and [here](https://github.com/aosp-mirror/platform_system_core/blob/android-12.1.0_r4/fs_mgr/include_fstab/fstab/fstab.h#L32)), assuming we don't drop it altogether. I have no idea what the implications are with changing from the `vendor_dlkm-verity` logical partition to `vendor_dlkm_a` or `vendor_dlkm_b`.

capntrips committed:
use official repo and associated binary

capntrips committed:
use replace instead of resize

capntrips commented:
I replaced the `resize` command with `replace` (see [here](https://github.com/phhusson/lp-self-flash/blob/d3890cae94f52bdd8afb5390ade1b52868f53cea/flash.sh#L15)), and it dramatically reduced the complexity, as `replace` doesn't require that `vendor_dlkm` be unmounted or unmapped.

The only downsides I can come up with are it requires available free space on the `super` partition to create the temporary partition, and nothing will be able to read from the new `vendor_dlkm` until after a reboot (or after the same unmount, unmap, and map that the `resize` method used).

It might make sense to add some error checking to the `replace` command.

capntrips committed:
added error checking to lptools and flash_generic

capntrips commented:
I added error checking to `lptools` (see PR [here](https://github.com/phhusson/vendor_lptools/pull/4)) and `flash_generic`. That's hopefully the last of the changes. I'll squash all the commits if no changes are needed.

osm0sis commented:
@engstk @freak07 can you guys vet these changes on your Pixel 6's to ensure we're not missing anything?

osm0sis commented:
@capntrips does lptools print any output if there's a fault? If so all those `if [ $? != 0 ];` would be unnecessarily.. also a simple || should suffice for all of those (if needed) to keep them concise. 👍

capntrips commented:
Yes, it [prints](https://github.com/phhusson/vendor_lptools/blob/ff09b3f150cb0bceaf7c3bd428bfec812d7440bc/lptools.cc#L131) a line at the start for all commands.

capntrips committed:
make error handling more concise

osm0sis commented:
Why does it reset imgblock using lptools map?

Edit: Oh, I see! Because it's $1_ak3 at that point which we need to write to instead, provided it maps. 👍

Should still be under /dev/block/mapper/$1_ak3 though shouldn't it? (Though I see that's not what phhusson did in lp-self-flash.. 🤔)

Edit 2: I think the "image larger than" check/abort needs to stay for non-logical partitions. So maybe doing that part as a case match with the abort as the fallback would make more sense.

Might be good to have the size check occur anyway and only resize vendor_dlkm from the stock size if larger is required too, to avoid messing with super unless absolutely necessary; so just put your new stuff under that old check. 🙂

Also please use == instead of = where still needed; it's not portable but it's an old Android shell convention and stylistically what I've used throughout AK3. 🙃

capntrips committed:
only replace logical partition if image is larger

use /dev/block/mapper, rather than /dev/block/dm-

capntrips commented:
I think it's safe to assume if the map operation doesn't fail, that `/dev/block/mapper/$1_ak3` will exist ([here](https://github.com/aosp-mirror/platform_system_core/blob/android-12.1.0_r4/init/devices.cpp#L347) seems to be where the symlink is made).

One scenario where always replacing `vendor_dlkm` would be beneficial, regardless of size, is if someone flashes two different kernels back to back without a reboot. For example, blu_spark doesn't modify `dtbo`, but Kirisakura does, so for testing previously, I was able to flash from Kirisakura back to stock (using AK3 with an `simg2img`'d `vendor_dlkm.img`) then immediately flash blu_spark without a reboot in between. After the last commit, performing that same operation will overwrite the old Kirisakura extents that will no longer be used on the next reboot. Note that blu_spark's `vendor_dlkm` is the smallest and the stock one is the largest of the three.

If anyone wants to test that the resize actually happens when needed will need to shrink their `vendor_dlkm` partition, first. Assuming you're already on blu_spark and slot `a`:

```sh
unzip blu_spark_r60-pxl6_cdc2dfef9.zip vendor_dlkm.img
lptools remove vendor_dlkm_ak3
lptools create vendor_dlkm_ak3 $(wc -c < vendor_dlkm.img)
lptools unmap vendor_dlkm_ak3
lptools map vendor_dlkm_ak3
dd if=vendor_dlkm.img of=/dev/block/mapper/vendor_dlkm_ak3 bs=1048576
lptools replace vendor_dlkm_ak3 vendor_dlkm_a
reboot
```

or flash `vendor_dlkm` from `fastbootd`.

capntrips committed:
missed an equality operator

osm0sis commented on tools/ak3-core.sh:
```diff
@@ -437,9 +437,14 @@ flash_generic() {
-      abort "New $1 image larger than $1 partition. Aborting...";
+      [ $path == /dev/block/mapper ] || abort "New $1 image larger than $1 partition. Aborting...";
+      islp=true;
+      $bin/lptools remove $1_ak3 || abort "Removing $1_ak3 failed. Aborting...";
```
Won't this abort always trigger since (unless something has failed previously) $1_ak3 should never exist?

capntrips replied:
The exit status is 0, even if the partition doesn't exist. The only thing that causes it to return a different exit status is if it fails to write the partition table after the operation (which, unless I'm missing something, would be totally unrelated to the remove operation). This is consistent with how it is handled in `fastboot`/`fs_mgr`.

osm0sis replied:
Hmm fair enough, but if there's nothing to remove then shouldn't there be nothing to change in the partition table? Just curious/confused.

Why bother with the abort here then if it wouldn't actually be failing on the removal?

capntrips replied:
The `replace` operation also removes the source partition, so it should never exist at that point. I assume it was there in `lp-self-flash` to clean up in the case of a failure after the initial `create` that prevented `replace` from running.

**Edit:** I only included the abort since there is at least the possibility that the operation will fail, if however unlikely.

osm0sis replied:
Right, so the remove line should probably stay in case that could happen for AK3 as well, but the || abort should probably be removed on that line since it's sounding superfluous; any potential table writing issue will still get caught by the create || abort line.

osm0sis reviewed (changes requested):
If this is a static build of lptools, as you say, then shouldn't it end up as lptools_static?

https://github.com/phhusson/vendor_lptools/blob/master/Android.bp#L32

Also please see my above line comment. 👍

capntrips commented:
I renamed the binary after building. I assumed the only reason it is named differently during build is soong doesn't allow multiple modules with the same name. I'm happy to change it back, if you would prefer.

osm0sis commented:
> I renamed the binary after building. I assumed the only reason it is named differently during build is soong doesn't allow multiple modules with the same name. I'm happy to change it back, if you would prefer.

Great. Yes, please! I'm a fan of having things named as built. 🙂

capntrips committed:
restore lptools_static name

remove superfluous error message

capntrips committed:
fix lptools_static binary name in script

capntrips commented:
I got ahead of myself with pushing the commits. I still need to run some tests verify everything still works after the changes.

**Edit:** Also, I can't find the button to resolve the change to the binary name.

osm0sis commented:
I'm going to push some minor tweaks real quick too, then all looks good to me, but we'll just need @engstk or @freak07 to test it since I'd just like external verification it all works as intended/desired where I don't have a device to test it personally. 🤞

osm0sis committed:
Minor tweaks for consistency

osm0sis committed:
Touch up LICENSE

osm0sis commented:
Alrighty, all good on my side. 🙌

@capntrips Please verify all still works for you, then I'll try to rope in one of those guys. 😛

capntrips commented:
A few tests both resizing and writing a smaller image within a larger partition passed without issue.

Also, I didn't see a comment on the following, so I just wanted to note it again, in case it is an issue:

> One scenario where always replacing `vendor_dlkm` would be beneficial, regardless of size, is if someone flashes two different kernels back to back without a reboot. For example, blu_spark doesn't modify `dtbo`, but Kirisakura does, so for testing previously, I was able to flash from Kirisakura back to stock (using AK3 with an `simg2img`'d `vendor_dlkm.img`) then immediately flash blu_spark without a reboot in between. After the last commit, performing that same operation will overwrite the old Kirisakura extents that will no longer be used on the next reboot. Note that blu_spark's `vendor_dlkm` is the smallest and the stock one is the largest of the three.

The solution is to ensure users reboot between flashes, but I can see that instruction being easily overlooked.

osm0sis commented:
I don't see why that wouldn't still be fine.. the operations are performed on super in real time, they just aren't loaded into memory until a reboot. 🤔

capntrips commented:
Right, the issue only arises after `replace` followed by additional attempts to read the size of or write to `vendor_dlkm` prior to a reboot, as those operations will still reference the old partition, rather than the new.

engstk commented:
I think it should have always replacing,which is the default behavior for other partitions.
For example, even with same kernel zips, sometimes people just flash a previous build and only notice at flash time.
Ensure latest flash is the one we want and boot with that. Checking space of N state or N-1 state should be transparent in the event of eventual resizing (since lower is "safe").
Replace should be "dumb", always do the work. Trusting a reboot is no bueno.

engstk commented:
All seems ok from my testing

osm0sis commented:
> Right, the issue only arises after `replace` followed by additional attempts to read the size of or write to `vendor_dlkm` prior to a reboot, as those operations will still reference the old partition, rather than the new.

> I think it should have always replacing,which is the default behavior for other partitions. For example, even with same kernel zips, sometimes people just flash a previous build and only notice at flash time. Ensure latest flash is the one we want and boot with that. Checking space of N state or N-1 state should be transparent in the event of eventual resizing (since lower is "safe"). Replace should be "dumb", always do the work. Trusting a reboot is no bueno.

Hmm, okay, so is there something that can be fixed here?

capntrips commented:
The simplest solution is to break the logical partition check out of the image larger than partition block and perform the replace operation every time. External tools still won't be able to access that data until after a reboot, but there won't be any unexpected data loss if running AK3 multiple times.

```diff
diff --git a/tools/ak3-core.sh b/tools/ak3-core.sh
index 4944976..e463d0f 100755
--- a/tools/ak3-core.sh
+++ b/tools/ak3-core.sh
@@ -437,14 +437,15 @@ flash_generic() {
     if [ ! "$imgblock" ]; then
       abort "$1 partition could not be found. Aborting...";
     fi;
-    if [ "$(wc -c < $img)" -gt "$(wc -c < $imgblock)" ]; then
-      [ "$path" == "/dev/block/mapper" ] || abort "New $1 image larger than $1 partition. Aborting...";
+    if [ "$path" == "/dev/block/mapper" ]; then
       islp=1;
       $bin/lptools_static remove $1_ak3;
       $bin/lptools_static create $1_ak3 $(wc -c < $img) || abort "Creating $1_ak3 failed. Aborting...";
       $bin/lptools_static unmap $1_ak3 || abort "Unmapping $1_ak3 failed. Aborting...";
       $bin/lptools_static map $1_ak3 || abort "Mapping $1_ak3 failed. Aborting...";
       imgblock=/dev/block/mapper/$1_ak3;
+    elif [ "$(wc -c < $img)" -gt "$(wc -c < $imgblock)" ]; then
+      abort "New $1 image larger than $1 partition. Aborting...";
     fi;
     isro=$(blockdev --getro $imgblock 2>/dev/null);
     blockdev --setrw $imgblock 2>/dev/null;
```

To make the data accessible without a reboot, the partition would have to be unmounted, unmapped, then remapped. I would think if someone needs that, they could perform those operations on their own.

engstk commented:
Yeah, I think it's ok and preferable that way.

osm0sis commented:
The img vs. partition check still needs to occur for all partition types though

Edit: Oh wait, nevermind. Re-read. 😛

osm0sis commented:
Okay, last question. Any reason we can't do the replace as part of the == /dev/block/mapper chunk and write the data to it after?

Like seems like it could go: remove _ak3, create _ak3, replace _ak3 $slot, unmap $slot, map $slot; then no need to change $imgblock ot set $islp, it keeps all the lp stuff together in the one if statement and the dd can just proceed afterward as with any other partition.

osm0sis committed:
Add proposed serial flashing fix

capntrips committed:
replace before flash

capntrips commented:
I ran some tests with `replace` before flash, and it worked without issue. As for unmapping then remapping `$1$slot`, that requires unmounting the filesystem; otherwise, `unmap` fails. If we're going to unmount and unmap `vendor_dlkm`, we may as well use `resize`, as with my [initial commit](https://github.com/osm0sis/AnyKernel3/pull/69/commits/c9f17255384369294ec0da7a29de35b736c814e9#diff-20ecbcf163865150c25972960038a9a8b9dccde3bc74d10976b52b38da0e681cR442-R470).

When verity is enabled, you potentially have to unmount  `$1-verity`, rather than `$1$slot`, and unmap them both. I say potentially, as after temporarily disabling verity for testing, I currently have verity enabled, but I only have the `$1$slot` partition. I've been meaning to do a clean flash to see if it comes back.

osm0sis commented:
Looks like it still flashes to the imgblock= _ak3 location after the latest commit? Not sure how that works with the replace before it... 😕🤔

capntrips commented:
Writing to `vendor_dlkm_ak3` after `replace` works for the same reason that you can't just write to `vendor_dlkm$slot`, at that point. Once the device is mapped, it points to the same location on disk until remapped, even if the underlying partition table changes. That's why a reboot would be required between flashes, if it were only conditionally replaced.

As for ditching `vendor_dlkm_ak3` altogether, that requires unmounting `vendor_dlkm`, which was the subject of the bulk of my previous comment. I know it can be done, as I had it working before switching to `replace`, after seeing how it was done in `lp-self-flash`. I'm happy to make the change, if you'd prefer to go that route.

osm0sis commented:
@engstk @freak07 any thoughts on that, or just merge as is? Please give it another quick test in its final state, when we're there.

@phhusson, if you've got a minute, was replace only used in lp-self-flash before resize was written, or is there anything more we should be aware of which would make us want to use replace over resize?

freak07 commented:
> @engstk @freak07 any thoughts on that, or just merge as is? Please give it another quick test in its final state, when we're there.

I´ll try to find time once it´s in the final state. Please ping me again. 

Same as @engstk stated earlier. It should not rely on a reboot in between to eliminate a potential source of user errors.

osm0sis commented:
The reboot concern should be resolved now. Now it's just back to the question of resize vs. replace and why phhusson used replace in lp-self-flash (if resize was an option at the time). 👍

phhusson commented:
I used replace because I wanted to replace system while system was up and running, so it needs to be non-conflicting extents, so it first need to allocate a new partition to get new extents then replace the ole system with the new one

osm0sis commented:
Perfect, thanks for explaining!🤘

So that's what we'll be allowing for vendor_dlkm (and wherever else this ends up being used) with AK3 booted flashing supported, so definitely the current setup with replace is the right way. 🎉

@engstk @freak07 sounds like it's ready for the final once-over!  🙌

freak07 commented:
looks like it´s working fine on my end. 

flashed kernel.zip which contains "regular" sized vendor_dlkm via updated ak3.zip and rebooted.

now flashed kernel.zip which contains "smaller" sized vendor_dlkm with the unused wifi driver disabled (about 6mb smaller) via updated ak3.zip and rebooted.

attempted to flash old ak3.zip, with "regular" sized vendor_dlkm on top of the new smaller vendor_dlkm, aborts as expected. 

attempted to flash updated ak3.zip with the regular sized vendor_dlkm, succeeds and boots fine. 


nice work @capntrips and @osm0sis!

engstk reviewed (approved):
All seems ok from me, it's a go.

osm0sis committed:
Bump core ver

osm0sis committed:
Merge branch 'master' into logical-resize

osm0sis commented:
Done! 🎉 

Thanks @capntrips for taking on the TODOs I left in the code for this important feature, since I don't have a vendor_dlkm device to test with yet! Very much appreciated! And as usual thanks to @engstk and @freak07 for testing it all out and being on the bleeding edge of AK3 all the time for their kernels. 😁

Now we just need someone to add vendor_boot v4 bootconfig and mult-ramdisk support to magiskboot... 😜

capntrips commented:
~After upgrading to the May update, I have `vendor_dlkm-verity`, rather than `vendor_dlkm_a`, which makes it choke on the [imgblock check](https://github.com/osm0sis/AnyKernel3/blob/79cb6945ff2eb0daf76d713ab466c95e71fc1743/tools/ak3-core.sh#L437).~

~I added `$1-verity` to the list and got it installed, but I'm havijg the same bootloop issue that I had with Magisk 24306 in April, so I can't yet confirm that it works as expected.~

**Edit:** I have been unable to replicate it. I'll wipe my data shortly, see if that triggers it.
**Edit 2:** This was likely user error. I have the ability to flash to the inactive slot to my app, and I must have clicked slot A by accident. I guess I should add some warnings.
